src/Components/Server/src/ComponentsHub.cs

```diff
@@ -17,8 +18,11 @@ namespace Microsoft.AspNetCore.Components.Server
     /// </summary>
     public sealed class ComponentsHub : Hub
     {
+        private static readonly TimeSpan LockAcquisitionTimeout = TimeSpan.FromSeconds(10);
+        private static readonly SemaphoreSlim CircuitRegistryLock = new SemaphoreSlim(1);
```

Comment:
Is it a better tradeoff to put these things in the registry and make those methods async? It seems like it would be a lot easier to test the various state transitions that way.